cmake/multi_image.cmake

```diff
@@ -173,7 +173,130 @@ function(get_child_image_cmake_args child_image output_cmake_args)
 
   unset(${output_cmake_args} PARENT_SCOPE)
   set(${output_cmake_args} "${child_image_cmake_args}" PARENT_SCOPE)
+endfunction()
 
+# Adds an S1 variant of an image to the build.
+#
+# Required arguments are:
+# SOURCE_DIR - The source dir of the base image.
+#
+# Optional argumnets are:
+# BASE_IMAGE - The name of the base image which the variant image is based on.
+function(add_s1_variant)
+  set(oneValueArgs BASE_IMAGE SOURCE_DIR)
+  cmake_parse_arguments(ASV "" "${oneValueArgs}" "" ${ARGN})
+
+  if (NOT ASV_SOURCE_DIR)
+    message(FATAL_ERROR "Missing parameter, required: SOURCE_DIR")
+  endif()
+
+  set(s1_override_conf
+    ${ZEPHYR_NRF_MODULE_DIR}/subsys/bootloader/image/is_s1_image.conf)
+
+  add_image_variant(
+    IMAGE_NAME s1_image
+    BASE_IMAGE ${ASV_BASE_IMAGE}
+    EXTRA_CONFIG ${s1_override_conf}
+    SOURCE_DIR ${ASV_SOURCE_DIR}
+    )
+endfunction()
+
+# Adds a variant of an image to the build.
+# The configuration passed to the variant image is a copy of that sent to the
+# base image, except what is passed to the EXTRA_CONFIG argument.
+#
+# Required arguments are:
+# IMAGE_NAME - The name of the variant image
+# SOURCE_DIR - The source dir of the base image.
+# EXTRA_CONFIG - The extra configuration fragment to pass to the variant image.
+#
+# Optional argumnets are:
+# BASE_IMAGE - The name of the base image which the variant image is based on.
+function(add_image_variant)
+
+  set(oneValueArgs BASE_IMAGE IMAGE_NAME SOURCE_DIR EXTRA_CONFIG)
+  cmake_parse_arguments(AIV "" "${oneValueArgs}" "" ${ARGN})
+
+  if (NOT AIV_IMAGE_NAME OR NOT AIV_SOURCE_DIR OR NOT AIV_EXTRA_CONFIG)
+    message(FATAL_ERROR "Missing parameter, required: IMAGE_NAME SOURCE_DIR EXTRA_CONFIG ")
+  endif()
+
+  if (AIV_BASE_IMAGE)
+    # A base image has been specified, this means that we need to copy the
+    # configuration of a child image (as opposed to copying the configuration
+    # from the top level parent image (that is, 'app').
+
+    set(base_image_configuration "${${AIV_BASE_IMAGE}_IMAGE_CMAKE_ARGS}")
+  else()
```

Comment:
if we use properties on the child image target and then create a pseudo target for the parent image with its settings, then we can treat the parent image like any other image and create a variant of it.

No need for having the variant code itself doing the differentiation.